[1.20.6] Use stack hash strategy for mutable map in creative tab event

dhyces opened the pull request:
Vanilla disallows duplicate elements in creative mode tabs, throwing a pretty unhelpful error:
```
Accidentally adding the same item stack twice [Wooden Sword] to a Creative Mode Tab: Damaged Wooden Swords
```
The current `BuildCreativeModeTabContentsEvent` mutable entries map does allow duplicate elements due to using default object comparison and hashing. This PR changes the hash strategy used to `ItemStackLinkedSet#TYPE_AND_TAG`

Fixes #1040.

dhyces committed:
Use stack hash strategy

dhyces committed:
Add test

TelepathicGrunt commented:
Uhhh. Are you _positive_ you are not going to reintroduce this bug where the creative menu enchanted books gets broken or order incorrectly? https://github.com/neoforged/NeoForge/pull/920

I feel like there was some issue I had before with the hashing and enchanted books in the menu. And had to remove the hashing in order to allow the enchanted books to show up properly

dhyces commented:
I was not aware of that issue. This does reintroduce that bug. If this does not have a proper solution, then a better error should at least be implemented along with an explanation on the event hook with why the custom hash strategy cannot be used.

TelepathicGrunt commented:
If testing and the enchanted book bug doesn’t happen, then this pr should be alright. Though maybe we should have a junit test added to verify that two of exact same item errors but enchanted books passes. See my Neo startup config PR for an example of a junit test that is also a test mod

TelepathicGrunt commented:
ah I see how vanilla bypasses their enchanted books. They deliberately skip checking search only entries lmao. 
<img width="808" alt="image" src="https://github.com/neoforged/NeoForge/assets/40846040/ee47a9b3-b3dd-4d90-a8ca-b1505ab1ede2">

Tested PR and it does break the books tho. While the enchanted books do show up in the search, the normal books are not showing under the Ingredients tabs.
<img width="929" alt="image" src="https://github.com/neoforged/NeoForge/assets/40846040/cd73c81b-3b4d-4daf-bce3-4c9405f63d73">

I think this PR's fix needs to change to allow duplicate itemstacks if the duplicates are search only mode. Also note, vanilla's check is actually order dependent too! So if you add search only itemstack first, and then the normal tab itemstack, you hit the duplicate exception error. But not the other way around. So that check will also need to be added to neo

TelepathicGrunt reviewed (changes requested):
Ok lets see. I need to leave a proper review asking for changes. PR just needs to solve this bug:
<img width="929" alt="image" src="https://github.com/neoforged/NeoForge/assets/40846040/442713fa-ce3e-49b2-8345-9d2504f89db9">

dhyces committed:
Write up unit test for creative tab enchantments

dhyces committed:
RFC basically

Not a big fan of the impl. Honestly would rather just remove the map entirely and use something else

dhyces committed:
Apply formatting to jdocs
